### PR TITLE
Add better bracket validation for Brain****; Fixes #625

### LINKED ISSUE #625: Brain**** bracket matching could be better
The current bracket matching code only counts the number of `[` and `]`. An invalid BF submission can pass this check, but it will receive a weird compile error message.

For example, this BF submission: `++++++]-----[++++`
will receive this compile error:
```
tletestc.c:23:36: error: expected identifier or ‘(’ before ‘--’ token
     ++*p;++*p;++*p;++*p;++*p;++*p;}--*p;--*p;--*p;--*p;--*p;while(*p){++*p;++*p;++*p;++*p;
                                    ^~
tletestc.c:23:41: error: expected identifier or ‘(’ before ‘--’ token
     ++*p;++*p;++*p;++*p;++*p;++*p;}--*p;--*p;--*p;--*p;--*p;while(*p){++*p;++*p;++*p;++*p;
                                         ^~
tletestc.c:23:46: error: expected identifier or ‘(’ before ‘--’ token
     ++*p;++*p;++*p;++*p;++*p;++*p;}--*p;--*p;--*p;--*p;--*p;while(*p){++*p;++*p;++*p;++*p;
                                              ^~
tletestc.c:23:51: error: expected identifier or ‘(’ before ‘--’ token
     ++*p;++*p;++*p;++*p;++*p;++*p;}--*p;--*p;--*p;--*p;--*p;while(*p){++*p;++*p;++*p;++*p;
                                                   ^~
tletestc.c:23:56: error: expected identifier or ‘(’ before ‘--’ token
     ++*p;++*p;++*p;++*p;++*p;++*p;}--*p;--*p;--*p;--*p;--*p;while(*p){++*p;++*p;++*p;++*p;
                                                        ^~
tletestc.c:23:61: error: expected identifier or ‘(’ before ‘while’
     ++*p;++*p;++*p;++*p;++*p;++*p;}--*p;--*p;--*p;--*p;--*p;while(*p){++*p;++*p;++*p;++*p;
                                                             ^~~~~
```

The expected error message is: `Unmatched brackets`

### PR DESCRIPTION
**Pull request by Carson-Tang**

### CODE REVIEW

File: dmoj/executors/BF.py [hunk 1]
```diff
@@ -58,6 +58,19 @@ def launch(self, *args, **kwargs):
...
 
+    def invalid_brackets(self, source_code) -> bool:
+        open_brackets = 0
+        valid = True
+        for c in source_code:
+            if c == ord(b'['):
+                open_brackets += 1
+            if c == ord(b']'):
+                if open_brackets == 0:
+                    valid = False
```
Comment (Xyene) [inline dmoj/executors/BF.py]: `return True` immediately here

File: dmoj/executors/BF.py [hunk 2]
```diff
@@ -58,6 +58,19 @@ def launch(self, *args, **kwargs):
         kwargs['memory'] += 8192
         return super().launch(str(memory * 1024), **kwargs)
 
+    def invalid_brackets(self, source_code) -> bool:
+        open_brackets = 0
+        valid = True
+        for c in source_code:
```
Comment (Xyene) [inline dmoj/executors/BF.py]: Leave a blank line before this.

File: dmoj/executors/BF.py [hunk 3]
```diff
@@ -58,6 +58,19 @@ def launch(self, *args, **kwargs):
         kwargs['memory'] += 8192
         return super().launch(str(memory * 1024), **kwargs)
 
+    def invalid_brackets(self, source_code) -> bool:
```
Comment (Xyene) [inline dmoj/executors/BF.py]: Either invert this to `_are_brackets_balanced`, or prefix with `_has_`.

File: dmoj/executors/BF.py [hunk 4]
```diff
@@ -58,6 +58,19 @@ def launch(self, *args, **kwargs):
...
+        open_brackets = 0
+        valid = True
+        for c in source_code:
+            if c == ord(b'['):
+                open_brackets += 1
+            if c == ord(b']'):
```
Comment (Xyene) [inline dmoj/executors/BF.py]: `elif`?

File: dmoj/executors/BF.py [hunk 5]
```diff
@@ -58,6 +58,19 @@ def launch(self, *args, **kwargs):
         kwargs['memory'] += 8192
         return super().launch(str(memory * 1024), **kwargs)
 
+    def invalid_brackets(self, source_code) -> bool:
+        open_brackets = 0
+        valid = True
+        for c in source_code:
+            if c == ord(b'['):
```
Comment (quantum5) [inline dmoj/executors/BF.py]: Instead of `ord` simply do `[0]`.

**Pull request merged**